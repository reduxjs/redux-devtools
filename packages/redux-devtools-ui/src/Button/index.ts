export { default } from './Button.js';
