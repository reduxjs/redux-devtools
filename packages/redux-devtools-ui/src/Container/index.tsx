import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';
import { useTheme, ThemeData } from '../utils/theme';
import { MainContainerWrapper, ContainerWrapper } from './styles';
import { Theme } from '../themes/default';

interface ContainerFromThemeDataProps {
  children?: React.ReactNode;
  themeData: ThemeData;
  className?: string;
}

const ContainerFromThemeData: React.FunctionComponent<
  ContainerFromThemeDataProps
> = ({ themeData, className, children }) => {
  const theme = useTheme(themeData);
  return (
    <ThemeProvider theme={theme}>
      <MainContainerWrapper className={className}>
        {children}
      </MainContainerWrapper>
    </ThemeProvider>
  );
};

interface Props {
  children?: React.ReactNode;
  themeData?: ThemeData;
  theme?: Theme;
  className?: string;
}

const Container: React.FunctionComponent<Props> = ({
  themeData,
  className,
  theme,
  children,
}) => {
  if (!themeData) {
    return (
      <ContainerWrapper className={className} theme={theme}>
        {children}
      </ContainerWrapper>
    );
  }

  return (
    <ContainerFromThemeData themeData={themeData} className={className}>
      {children}
    </ContainerFromThemeData>
  );
};

Container.propTypes = {
  children: PropTypes.node,
  themeData: PropTypes.any,
  theme: PropTypes.any,
  className: PropTypes.string,
};

export default Container;
