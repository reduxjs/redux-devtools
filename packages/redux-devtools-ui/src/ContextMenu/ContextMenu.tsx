import React, { Component } from 'react';
import PropTypes from 'prop-types';
import createStyledComponent from '../utils/createStyledComponent';
import styles from './styles/index';

const ContextMenuWrapper = createStyledComponent(styles);

type ReactButtonElement = React.ReactElement<
  JSX.IntrinsicElements['button'],
  'button'
>;
type Item = { name: string; value?: string } | ReactButtonElement;

function isReactButtonElement(item: Item): item is ReactButtonElement {
  return (item as ReactButtonElement).type === 'button';
}

export interface ContextMenuProps {
  items: Item[];
  onClick: (value: string) => void;
  x: number;
  y: number;
  visible?: boolean;
}

export default class ContextMenu extends Component<ContextMenuProps> {
  menu?: HTMLDivElement | null;

  componentDidMount() {
    this.amendPosition();
  }

  componentDidUpdate(prevProps: ContextMenuProps) {
    if (prevProps.x !== this.props.x || prevProps.y !== this.props.y) {
      this.amendPosition();
    }
  }

  onMouseUp: React.MouseEventHandler<HTMLButtonElement> = (e) => {
    e.currentTarget.blur();
  };

  onClick: React.MouseEventHandler<HTMLButtonElement> = (e) => {
    this.props.onClick(e.currentTarget.value);
  };

  amendPosition() {
    const { x, y } = this.props;
    const { scrollTop, scrollLeft } = document.documentElement;
    const { innerWidth, innerHeight } = window;
    const rect = this.menu!.getBoundingClientRect();
    let left = x + scrollLeft;
    let top = y + scrollTop;

    if (y + rect.height > innerHeight) {
      top = innerHeight - rect.height;
    }
    if (x + rect.width > innerWidth) {
      left = innerWidth - rect.width;
    }
    if (top < 0) {
      top = rect.height < innerHeight ? (innerHeight - rect.height) / 2 : 0;
    }
    if (left < 0) {
      left = rect.width < innerWidth ? (innerWidth - rect.width) / 2 : 0;
    }

    this.menu!.style.top = `${top}px`;
    this.menu!.style.left = `${left}px`;
  }

  renderItems() {
    return this.props.items.map((item) => {
      if (isReactButtonElement(item)) return item;
      const value = item.value || item.name;
      return (
        <button
          key={value}
          value={value}
          onMouseUp={this.onMouseUp}
          onClick={this.onClick}
        >
          {item.name}
        </button>
      );
    });
  }

  menuRef: React.RefCallback<HTMLDivElement> = (c) => {
    this.menu = c;
  };

  render() {
    return (
      <ContextMenuWrapper
        ref={this.menuRef}
        left={this.props.x}
        top={this.props.y}
        visible={this.props.visible}
      >
        {this.renderItems()}
      </ContextMenuWrapper>
    );
  }

  static propTypes = {
    items: PropTypes.array.isRequired,
    onClick: PropTypes.func.isRequired,
    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    visible: PropTypes.bool,
  };
}
