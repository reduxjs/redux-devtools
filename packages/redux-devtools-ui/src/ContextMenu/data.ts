export const items = [
  {
    name: 'Menu Item 1',
  },
  {
    name: 'Menu Item 2',
  },
  {
    name: 'Menu Item 3',
  },
];
