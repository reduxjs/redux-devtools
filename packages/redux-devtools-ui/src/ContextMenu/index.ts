export { default } from './ContextMenu.js';
