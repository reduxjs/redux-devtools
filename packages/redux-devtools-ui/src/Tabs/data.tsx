import React from 'react';

/* eslint-disable react/prop-types */
const Component = ({ selected }: { selected: string }) => (
  <div
    style={{
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'center',
      width: '100%',
      height: '100%',
      fontSize: '22px',
    }}
  >
    Selected {selected}
  </div>
);
/* eslint-enable react/prop-types */

const selector = (tab: { name: string; value?: string }) => ({
  selected: tab.name,
});

export const tabs = [
  {
    name: 'Tab1',
    component: Component,
    selector,
  },
  {
    name: 'Tab2',
    component: Component,
    selector,
  },
  {
    name: 'Tab3',
    component: Component,
    selector,
  },
];

export const simple10Tabs: { name: string; value: string }[] = [];
for (let i = 1; i <= 10; i++)
  simple10Tabs.push({ name: `Tab${i}`, value: `${i}` });
