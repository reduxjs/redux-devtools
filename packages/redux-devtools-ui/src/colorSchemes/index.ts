export { default } from './default';
export { default as github } from './github';
export { default as atomOneDark } from './atom-one-dark';
export { default as dracula } from './dracula';
export { default as iRBlack } from './ir-black';
export { default as macintosh } from './macintosh';
export { default as materia } from './materia';
export { default as oceanicNext } from './oceanic-next';
export { default as phD } from './phd';
export { default as pico } from './pico';
export { default as solarFlare } from './solar-flare';
export { default as spacemacs } from './spacemacs';
export { default as unikitty } from './unikitty';
export { default as woodland } from './woodland';
