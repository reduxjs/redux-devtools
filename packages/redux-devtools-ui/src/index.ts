export { default as Container } from './Container';
export { default as Button } from './Button';
export { default as ContextMenu } from './ContextMenu';
export { default as Dialog } from './Dialog';
export { default as Editor } from './Editor';
export { default as Form } from './Form';
export { default as Select } from './Select';
export { default as Slider } from './Slider';
export { default as Tabs, Tab } from './Tabs';
export { default as SegmentedControl } from './SegmentedControl';
export { default as Notification } from './Notification';
export * from './Toolbar';

import color from './utils/color';
export const effects = { color };
export { default as createStyledComponent } from './utils/createStyledComponent';
export { Theme, ThemeFromProvider, Scheme } from './utils/theme';
