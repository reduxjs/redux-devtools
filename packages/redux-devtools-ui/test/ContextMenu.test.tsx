import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { ContextMenu } from '../src';
import { items } from '../src/ContextMenu/data';

describe('ContextMenu', function () {
  it('renders correctly', () => {
    const { container } = render(
      <ContextMenu
        items={items}
        onClick={() => {
          // noop
        }}
        x={100}
        y={100}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  it('should handle the click event', async () => {
    const onClick = jest.fn();
    render(
      <ContextMenu items={items} onClick={onClick} x={100} y={100} visible />
    );

    await userEvent.click(screen.getByRole('button', { name: 'Menu Item 1' }));
    expect(onClick).toHaveBeenCalled();
  });
});
