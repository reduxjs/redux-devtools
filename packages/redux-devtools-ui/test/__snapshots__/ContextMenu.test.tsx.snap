// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu renders correctly 1`] = `
<div
  class="sc-jTQDnj ATyGi"
  style="top: 100px; left: 100px;"
>
  <button
    value="Menu Item 1"
  >
    Menu Item 1
  </button>
  <button
    value="Menu Item 2"
  >
    Menu Item 2
  </button>
  <button
    value="Menu Item 3"
  >
    Menu Item 3
  </button>
</div>
`;
